### PR TITLE
Contract Deployment

### PR DESCRIPTION
**Pull request by yann300**
- Register to Registrar contract.
 - Package dapp (ressources + manifest).

### CODE REVIEW

**Comment by yann300** [comment]
TODO: 
 - Manage multiple content-type.
 - Remove test action and hardcoded registrar, urlhint contract address.
 - Manage subfolder in package.dapp.

File: mix/FileIo.cpp [hunk 1]
```diff
@@ -20,13 +20,23 @@
  * Ethereum IDE client.
  */
 
+#include <QJsonObject>
```
Comment (arkpar) [inline mix/FileIo.cpp]: Is this required?

File: mix/FileIo.cpp [hunk 2]
```diff
@@ -101,3 +111,76 @@ bool FileIo::fileExists(QString const& _url)
 	QFile file(url.path());
 	return file.exists();
 }
+
+QStringList FileIo::compress(QString const& _deploymentFolder)
```
Comment (arkpar) [inline mix/FileIo.cpp]: Misleading name. please rename to makePackage or something like this

File: mix/FileIo.cpp [hunk 3]
```diff
@@ -101,3 +111,76 @@ bool FileIo::fileExists(QString const& _url)
...
+		QFile qFile(item.filePath());
+		if (qFile.open(QIODevice::ReadOnly | QIODevice::Text))
+				k++;
+	}
+	str.appendList(k);
+
+	for (auto item: deployDir.entryInfoList(QDir::Files))
+	{
+		QFile qFile(item.filePath());
+		if (qFile.open(QIODevice::ReadOnly | QIODevice::Text))
```
Comment (arkpar) [inline mix/FileIo.cpp]: File must be read in binary mode

File: mix/FileIo.cpp [hunk 4]
```diff
@@ -101,3 +111,76 @@ bool FileIo::fileExists(QString const& _url)
...
+	dev::h256 h = dev::sha3(dapp);
+	//encrypt
+	KeyPair key(h);
+	Secp256k1 enc;
+	enc.encrypt(key.pub(), dapp);
+
+	QUrl url(_deploymentFolder + "package.dapp");
+	QFile compressed(url.path());
+	QByteArray qFileBytes((char*)dapp.data(), dapp.size());
+	if (compressed.open(QIODevice::WriteOnly | QIODevice::Text))
```
Comment (arkpar) [inline mix/FileIo.cpp]: File must be read in binary mode

File: mix/FileIo.cpp [hunk 5]
```diff
@@ -101,3 +111,76 @@ bool FileIo::fileExists(QString const& _url)
...
+
+	for (auto item: deployDir.entryInfoList(QDir::Files))
+	{
+		QFile qFile(item.filePath());
+		if (qFile.open(QIODevice::ReadOnly | QIODevice::Text))
+		{
+			QFileInfo i = QFileInfo(qFile.fileName());
+
+			QByteArray _a = qFile.readAll();
+			bytes b = bytes(_a.begin(), _a.end());
```
Comment (arkpar) [inline mix/FileIo.cpp]: You can use bytesConstRef to avoid extra file copy here

File: mix/FileIo.cpp [hunk 6]
```diff
@@ -101,3 +111,76 @@ bool FileIo::fileExists(QString const& _url)
...
+			str.append(b);
+
+			Json::Value f;
+			f["path"] = "/"; //TODO: Manage relative sub folder
+			f["file"] = "/" + i.fileName().toStdString();
+			f["contentType"] = "application/html"; //TODO: manage multiple content type
```
Comment (arkpar) [inline mix/FileIo.cpp]: text/html

File: mix/FileIo.cpp [hunk 7]
```diff
@@ -101,3 +111,76 @@ bool FileIo::fileExists(QString const& _url)
...
+		if (qFile.open(QIODevice::ReadOnly | QIODevice::Text))
+		{
+			QFileInfo i = QFileInfo(qFile.fileName());
+
+			QByteArray _a = qFile.readAll();
+			bytes b = bytes(_a.begin(), _a.end());
+			str.append(b);
+
+			Json::Value f;
+			f["path"] = "/"; //TODO: Manage relative sub folder
```
Comment (arkpar) [inline mix/FileIo.cpp]: "/" should be the path for index.html. All other files should default to filename

File: mix/FileIo.cpp [hunk 8]
```diff
@@ -101,3 +111,76 @@ bool FileIo::fileExists(QString const& _url)
...
+		}
+		qFile.close();
+	}
+
+	manifest["entries"] = entries;
+
+	std::stringstream jsonStr;
+	jsonStr << manifest;
+	QByteArray b =  QString::fromStdString(jsonStr.str()).toUtf8();
+	str.append(bytes(b.begin(), b.end()));
```
Comment (arkpar) [inline mix/FileIo.cpp]: Manifest must be the first item in RLP, not last

File: mix/qml/DeploymentDialog.qml [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+		RowLayout
+		{
+			anchors.bottom: parent.bottom
+			anchors.right: parent.right;
+			anchors.bottomMargin: 10
+			Button {
+				text: qsTr("Deploy on Ethereum");
+				tooltip: qsTr("Deploy the contract and Package ressources files.")
+				onClicked: {
+					if (applicationUrlEth.text === "")
```
Comment (arkpar) [inline mix/qml/DeploymentDialog.qml]: Don't we allow installing into the root reg?

File: mix/qml/DeploymentDialog.qml [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+		}
+
+		RowLayout
+		{
+			anchors.bottom: parent.bottom
+			anchors.right: parent.right;
+			anchors.bottomMargin: 10
+			Button {
+				text: qsTr("Deploy on Ethereum");
+				tooltip: qsTr("Deploy the contract and Package ressources files.")
```
Comment (arkpar) [inline mix/qml/DeploymentDialog.qml]: Deploy contract and package resources files.

File: mix/qml/DeploymentDialog.qml [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+			icon: StandardIcon.Warning
+		}
+
+		RowLayout
+		{
+			anchors.bottom: parent.bottom
+			anchors.right: parent.right;
+			anchors.bottomMargin: 10
+			Button {
+				text: qsTr("Deploy on Ethereum");
```
Comment (arkpar) [inline mix/qml/DeploymentDialog.qml]: Deploy to

File: mix/qml/DeploymentDialog.qml [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+			}
+
+			Button {
+				text: qsTr("Register Web Application");
+				tooltip: qsTr("Register hosted Web Application.")
+				onClicked: {
+					if (applicationUrlHttp.text === "" || deploymentDialog.packageHash === "")
+					{
+						deployDialog.title = text;
+						deployDialog.text = qsTr("Please provide the link where the ressources are stored and ensure the package is aleary built using the deployment step. ")
```
Comment (arkpar) [inline mix/qml/DeploymentDialog.qml]: resources

File: mix/qml/ProjectModel.qml [hunk 13]
```diff
@@ -91,15 +93,44 @@ Item {
...
 		title: qsTr("Project")
-		text: qsTr("This project has been already deployed to the network. Do you want to re-deploy it?")
-		standardButtons: StandardButton.Ok | StandardButton.Cancel
+		text:
+		{
+			if (Object.keys(projectModel.deploymentAddresses).length > 0)
+			{
+				redeploy = true
+				standardButtons = StandardButton.Ok | StandardButton.Reset | StandardButton.Abort;
+				return qsTr("This project has been already deployed to the network. Do you want to repackage the ressources only, or also reset the deployed contract to his initial state?")
```
Comment (arkpar) [inline mix/qml/ProjectModel.qml]: "resources", "its initial state"
Also, I don't think this is needed at all. We should just overwrite everything

File: mix/qml/js/ProjectModel.js [hunk 14]
```diff
@@ -369,7 +387,203 @@ function finalizeDeployment(deploymentId, addresses) {
...
+			rpcCall(requests, function (httpRequest, response) {
+				callBack();
+			});
+		});
+	}
+}
+
+function registerToUrlHint()
+{
+	deploymentStepChanged(qsTr("Registering application Ressources (" + deploymentDialog.applicationUrlHttp) + ") ...");
```
Comment (arkpar) [inline mix/qml/js/ProjectModel.js]: resources

File: mix/qml/ProjectModel.qml (hunk 13 above)
Comment (yann300) [inline mix/qml/ProjectModel.qml]: The developper may have made some modifications on ressources (html/js...) but not on the contract. So in that case we don't need to redeploy the contract.

Comment (arkpar) [inline mix/qml/ProjectModel.qml]: Ok, I see the point now

**Pull request merged**